File: lib/core/transformers/standard/encoder_decoder_transformer.dart
/// Full encoder-decoder Transformer.
import 'package:dart_cuda/core/layers/nn.dart';
import 'package:dart_cuda/core/tensor/gpu_tensor.dart';
import 'package:dart_cuda/core/transformers/standard/transformer_decoder.dart';
import 'package:dart_cuda/core/transformers/standard/transformer_encoder.dart';

class EncoderDecoderTransformer extends Module {
  final TransformerEncoder encoder;
  final TransformerDecoder decoder;

  EncoderDecoderTransformer({
    required int sourceVocabSize,
    required int targetVocabSize,
    required int embedSize,
    required int sourceBlockSize,
    required int targetBlockSize,
    required int numLayers,
    required int numHeads,
  }) : encoder = TransformerEncoder(
         vocabSize: sourceVocabSize,
         embedSize: embedSize,
         blockSize: sourceBlockSize,
         numLayers: numLayers,
         numHeads: numHeads,
       ),
       decoder = TransformerDecoder(
         vocabSize: targetVocabSize,
         embedSize: embedSize,
         blockSize: targetBlockSize,
         numLayers: numLayers,
         numHeads: numHeads,
         encoderEmbedSize: embedSize,
       );

  Tensor forward(
    List<int> sourceIdx,
    List<int> targetIdx,
    List<Tensor> tracker,
  ) {
    final encoderOutput = encoder.forward(sourceIdx, tracker);
    return decoder.forward(targetIdx, encoderOutput, tracker);
  }

  @override
  List<Tensor> parameters() => [
    ...encoder.parameters(),
    ...decoder.parameters(),
  ];
}

File: python/dart_cuda/core/transformers/standard/encoder_decoder_transformer.py
"""Full encoder-decoder Transformer — port of
`lib/transformer/encoder_decoder_transformer.dart` (dart-torch).
"""

from __future__ import annotations

from typing import List, Sequence

from ...layers.nn import Module
from ...tensor.gpu_tensor import Tensor
from .transformer_encoder import TransformerEncoder
from .transformer_decoder import TransformerDecoder


class EncoderDecoderTransformer(Module):
    def __init__(
        self,
        *,
        source_vocab_size: int,
        target_vocab_size: int,
        embed_size: int,
        source_block_size: int,
        target_block_size: int,
        num_layers: int,
        num_heads: int,
    ):
        self.encoder = TransformerEncoder(
            vocab_size=source_vocab_size,
            embed_size=embed_size,
            block_size=source_block_size,
            num_layers=num_layers,
            num_heads=num_heads,
        )
        self.decoder = TransformerDecoder(
            vocab_size=target_vocab_size,
            embed_size=embed_size,
            block_size=target_block_size,
            num_layers=num_layers,
            num_heads=num_heads,
            encoder_embed_size=embed_size,
        )

    def forward(
        self,
        source_idx: Sequence[int],
        target_idx: Sequence[int],
        tracker: List[Tensor],
    ) -> Tensor:
        encoder_output = self.encoder.forward(source_idx, tracker)
        logits = self.decoder.forward(target_idx, encoder_output, tracker)
        return logits

    def parameters(self) -> List[Tensor]:
        return [*self.encoder.parameters(), *self.decoder.parameters()]
